### ISSUE TITLE
Using different scopes to inject steps into other steps

### ISSUE THREAD

**Issue by adhesivee** [cukespace/cukespace#91]
I am trying to use inject background steps into the "normal" steps. Without scopes it is not working, when I add the `@ApplicationScoped` it is still not working. The problem that I got is that at the time that it reaches a field with `@Inject` a different instance is injected then the one that is used for testing.

This is what I have, my test file:
```
@RunWith(ArquillianCucumber.class)
@CucumberOptions(features = "src/test/resources/features/")
public class FeatureTest {
}
```

Background steps:
```
@ApplicationScoped
public class BackgroundSteps {
    private String givenOutput;
    @Given("^...$")
    public void given() {
        givenOutput = "test";
    }

    public String getGivenOutput() {
        return givenOutput;
    }
}
```

Feature steps:
```
@ApplicationScoped
public class FeatureSteps {
    @Inject
    private BackgroundSteps steps;

    @Given("^...$")
    public void testSomeMethod() {
        System.out.println(steps.getGivenOutput());
    }
}
```
The output of this script is `null`.
I have validated that the steps runs in order. When I test this without arquillian and `cucumber-weld` this works. But not with this extension. When I validate the object that is used to run the `@Given` on and the `@Inject` when I see that they are to different instances:
The one from `@Given`: `BackgroundSteps@12142bc6`
The one from `@Inject`: `BackgroundSteps@d13c0e8`

How can I make it possible to inject other steps into each other?

**Comment by rmannibucau** [comment]
You need to use a scoped bean shared in both steps to store what you need:


    @ApplicationScoped
    public class SharedData {
        private String givenOutput; // setter/getter
    }
    
    public class BackgroundSteps {
        @inject SharedData data;
    
        @Given("^...$")
        public void given() {
            data.setGivenOutput("test");
        }
    }

    public class FeatureSteps {
        @Inject
        private SharedData data;
    
        @Given("^...$")
        public void testSomeMethod() {
            System.out.println(data.getGivenOutput());
        }
    }

**Comment by adhesivee** [comment]
I am using this at the moment, but I consider this more of a workaround. Because it is possible in the "plain" cucumber with weld. 
But I am not sure if it is possible with arquillian. Maybe the solution I am looking for shouldn't be fixed with this project. But do you have an idea how I can fix this, at least for myself? Currently I am looking into `CDIInjectionEnricher` to see if I can influence it somehow.

**Comment by rmannibucau** [comment]
Ok, this is a bit more vicious actually. TestEnrichers are to inject but not to create the instances. Instances (steps) are created by cucumber "backend". Cukespace uses https://github.com/cukespace/cukespace/blob/master/core/src/main/java/cucumber/runtime/arquillian/backend/ArquillianBackend.java which doesnt do a CDI lookup (obviously since it works with any container including spring, EE 5 - no cdi, plain tomcat etc...). In this class you can check the initInstances() method which basically just do a glueClass.newInstance(). So here is what is happening:

1. the backend ensures you get a single instance of each step
2. the backend does *not* add the steps in any context (CDI, spring, ...) cause no assumption is done there
3. if you inject a step you inject a bean which is not read from the backend

Weld implementation replaces the glue/step instance lookup by a CDI lookup which works but also enforces all steps to be CDI beans (which is rare actually).

Concretely today cukespace ignores cucumber.api.java.ObjectFactory

I added the support in arquillian.xml of `objectFactory` property if you want to use another one. You can set it to "cdi" for now to have a kind of behavior like the weld one.

Side note: you can't reuse weld ObjectFactory since the impl of cucumber-jvm also starts/stop weld itself which is done by arquillian with cukespace so it would conflict.

I deployed the snapshot on sonatype snapshot repository if you want to give it a try (1.6.7-SNAPSHOT).

**Comment by adhesivee** [comment]
Great! Thanks for the change. At the moment I get messages on every `@Given`, `@When` and `@Then` that it is not defined. Maybe I configured something wrong, but I will look into it later.

**Comment by adhesivee** [comment]
Any idea when this is going to be released? I can work with a local snapshot for now.

**Comment by rmannibucau** [comment]
just released 1.6.7
